decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,361 @@
+```
+
+### An example of using these specifications
+
+#### Checking management policies
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+imageSpec = ImageSpecification()
+policy = manager.managementPolicy([imageSpec], context)[0]
+
+if policy == [kIgnored]:
+    return  # nothing with an ImageSpecification is managed
+
+# check if pngs and alembic caches are supported by the manager
+exrSpec = ImageSpecification()
+exrSpec.mimeType = "image/x-exr; version=\"2\""
+
+abcSpec = SceneCacheSpecification()
+abcSpec.mimeType = "alembic"
+policy = manager.managementPolicy([exrSpec, abcSpec], context)[0]
+
+policy == [kManaged, kManaged]  # manager offers to manage these exrs and alembic
+```
+
+#### Finding all the caches associated with a shot - (such as to schedule for deletion to free up space)
+
+```python
+
+assets = ["asset://fxPointsCache",  # a geometry cache of point data
+          "asset://precompShadows"]  # rendered image of shadows
+
+shots = ["shot://ABC/01"]
+
+
+class CacheTrait(specifications.EntityTrait):
+    _type = "cache"
+
+
+class ImageTrait(specifications.EntityTrait):
+    _type = "image"
+
+
+class ImageCacheSpec(ImageTrait, CacheTrait):
+    _type = ".".join(sorted((ImageTrait._type, CacheTrait)))
+
+
+class GeometryTrait(specifications.EntityTrait):
+    _type = "geometry"
+
+
+class GeometryCacheSpec(GeometryTrait, CacheTrait):
+    _type = ".".join(sorted((GeometryTrait._type, CacheTrait)))
+
+
+class Bundle(specifications.RelationshipSpecification):
+    _type = "bundle"
+
+
+manager.getRelatedReferences(shots, Bundle, context, session,
+                             resultSpec=[ImageCacheSpec, GeometryCacheSpec])
+
+# expected output: [["asset://fxPointsCache"], ["asset://precompShadows"]]
+```
+
+### A possible example of using traits directly
+
+#### Checking if the manager manages specific extensions
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+imageTrait = ImageTrait()
+policy = manager.managementPolicy([imageTrait], context)[0]
+
+if policy == [kIgnored]:
+    return  # nothing with an image trait is managed
+
+# check if only a certain extension is supported by the manager
+blobTrait1 = BlobTrait()
+blobTrait.mimeType = "image/x-exr; version=\"2\""
+
+blobTrait2 = BlobTrait()
+blobTrait2.mimeType = "alembic"
+
+policy = manager.managementPolicy([blobTrait, blobTrait2], context)[0]
+
+policy == [kManaged, kManaged]  # manager offers to manage these exrs and alembic
+```
+
+This would require the manager interface to take a list of trait parameters instead of a list of
+specification parameters.
+
+#### Finding all the caches associated with a shot, using traits directly
+
+```python
+
+assets = ["asset://fxPointsCache",  # a geometry cache of point data
+          "asset://precompShadows"]  # rendered image of shadows
+
+shots = ["shot://ABC/01"]
+
+
+class CacheTrait(specifications.EntityTrait):
```

Comment:
[consider] Can we de-dupe the declarations of these traits to a section at the top? Then the examples can be a bit closer to gather, which makes it easier to read.

Reply:
Reorganised the declarations at the top 78c87c0d7038ae6c2f948e0d7ed5bc39023d108f